File: packages/typescript/src/types/components.ts
import type { Memory } from "./memory";
import type { Content } from "./primitives";
import type {
	JsonValue,
	ActionExample as ProtoActionExample,
	ActionParameter as ProtoActionParameter,
	ActionParameterSchema as ProtoActionParameterSchema,
	ActionParameters as ProtoActionParametersType,
	EvaluationExample as ProtoEvaluationExample,
} from "./proto.js";
import type { IAgentRuntime } from "./runtime";
import type { ActionPlan, State } from "./state";

/**
 * Canonical domain contexts for routing and plugin/action gating.
 *
 * The shouldRespond + context-routing classifier assigns one primary context
 * and zero or more secondary contexts per turn.  Plugins, actions, and
 * providers declare which contexts they belong to so the planner can scope
 * its search space accordingly.
 */
export type AgentContext =
	| "general"
	| "wallet"
	| "knowledge"
	| "browser"
	| "code"
	| "media"
	| "automation"
	| "social"
	| "system"
	| (string & {}); // extensible — plugins can declare custom contexts

/**
 * JSON Schema type for action parameter validation.
 * Supports basic JSON Schema properties for parameter definition.
 */
export interface ActionParameterSchema
	extends Omit<
		ProtoActionParameterSchema,
		| "$typeName"
		| "$unknown"
		| "defaultValue"
		| "properties"
		| "items"
		| "enumValues"
	> {
	/** Default value if parameter is not provided */
	default?: JsonValue | null;
	/** For object types, define nested properties */
	properties?: Record<string, ActionParameterSchema>;
	/** For array types, define the item schema */
	items?: ActionParameterSchema;
	/** Enumerated allowed values (schema-compatible) */
	enumValues?: string[];
	/** Enumerated allowed values */
	enum?: string[];
}

/**
 * Defines a single parameter for an action.
 * Parameters are extracted from the conversation by the LLM and passed to the action handler.
 */
export interface ActionParameter
	extends Omit<ProtoActionParameter, "$typeName" | "$unknown" | "schema"> {
	/** Parameter name (used as the key in the parameters object) */
	name: string;
	/** Human-readable description for LLM guidance */
	description: string;
	/** Compressed description for prompt-optimized rendering */
	descriptionCompressed?: string;
	/** Whether this parameter is required (default: false) */
	required?: boolean;
	/** JSON Schema for parameter validation */
	schema: ActionParameterSchema;
	/**
	 * Optional example values for this parameter.
	 * These are shown to the model in action descriptions to improve extraction accuracy.
	 */
	examples?: ActionParameterExampleValue[];
}

/**
 * Primitive value types that can be used in action parameters.
 */
export type ActionParameterValue = string | number | boolean | null;

/**
 * Example value types allowed for action parameter examples.
 * Supports primitives as well as nested objects/arrays for documentation purposes.
 */
export type ActionParameterExampleValue =
	| ActionParameterValue
	| ActionParameters
	| ActionParameterValue[]
	| ActionParameters[];

/**
 * Validated parameters passed to an action handler.
 * Keys are parameter names, values are the validated parameter values.
 * Supports nested objects and arrays for complex parameter structures.
 */
export interface ActionParameters {
	[key: string]:
		| ActionParameterValue
		| ActionParameters
		| ActionParameterValue[]
		| ActionParameters[]
		| JsonValue;
}

export type ProtoActionParameters = ProtoActionParametersType;

/**
 * Example content with associated user for demonstration purposes
 */
export interface ActionExample
	extends Omit<ProtoActionExample, "$typeName" | "$unknown" | "content"> {
	content: Content;
}

export interface EvaluationExample
	extends Omit<ProtoEvaluationExample, "$typeName" | "$unknown" | "messages"> {
	messages: ActionExample[];
}

/**
 * Callback function type for handlers. actionName is optional so callers can attribute
 * the response to the action that produced it without parsing content (backward compatible).
 */
export type HandlerCallback = (
	response: Content,
	actionName?: string,
) => Promise<Memory[]>;

/**
 * Handler function type for processing messages
 */
export type Handler = (
	runtime: IAgentRuntime,
	message: Memory,
	state?: State,
	options?: HandlerOptions | Record<string, JsonValue | undefined>,
	callback?: HandlerCallback,
	responses?: Memory[],
) => Promise<ActionResult | undefined>;

/**
 * Validator function type for actions/evaluators
 */
export type Validator = (
	runtime: IAgentRuntime,
	message: Memory,
	state?: State,
) => Promise<boolean>;

/**
 * Represents an action the agent can perform
 */
export interface Action {
	/** Action name */
	name: string;

	/** Detailed description */
	description: string;

	/** Compressed description for prompt-optimized action selection */
	descriptionCompressed?: string;

	/** Handler function */
	handler: Handler;

	/** Validation function */
	validate: Validator;

	/** Similar action descriptions */
	similes?: string[];

	/** Example usages */
	examples?: ActionExample[][];

	/** Optional priority for action ordering */
	priority?: number;

	/** Optional tags for categorization */
	tags?: string[];

	/**
	 * When true, the message service should stop after executing this action
	 * instead of running a post-action continuation LLM turn.
	 *
	 * Use this for actions that already emit a complete user-facing reply or
	 * that launch asynchronous background work whose progress will continue
	 * outside the current chat turn.
	 */
	suppressPostActionContinuation?: boolean;

	/**
	 * Optional input parameters for the action.
	 * When defined, the LLM will be prompted to extract these parameters from the conversation
	 * and they will be validated before being passed to the handler via HandlerOptions.parameters.
	 *
	 * Parameters can be required or optional. Optional parameters may have defaults
	 * or can be backfilled inside the action handler if not provided.
	 *
	 * @example
	 * ```typescript
	 * parameters: [
	 *   {
	 *     name: "targetUser",
	 *     description: "The username or ID of the user to send the message to",
	 *     required: true,
	 *     schema: { type: "string" }
	 *   },
	 *   {
	 *     name: "platform",
	 *     description: "The platform to send the message on (telegram, discord, etc)",
	 *     required: false,
	 *     schema: { type: "string", enum: ["telegram", "discord", "x"], default: "telegram" }
	 *   }
	 * ]
	 * ```
	 */
	parameters?: ActionParameter[];

	/**
	 * Domain contexts this action belongs to.
	 * Used by the context-routing classifier to scope the planner's action search.
	 * An action may belong to multiple contexts (e.g., a token-swap action is both
	 * "wallet" and "automation").
	 */
	contexts?: AgentContext[];
}

/**
 * Evaluator for assessing agent responses
 */
export interface Evaluator {
	/** Whether to always run */
	alwaysRun?: boolean;

	/** Detailed description */
	description: string;

	/** Similar evaluator descriptions */
	similes?: string[];

	/** Example evaluations */
	examples: EvaluationExample[];

	/** Handler function */
	handler: Handler;

	/** Evaluator name */
	name: string;

	/** Validation function */
	validate: Validator;
}

/**
 * JSON-serializable primitive values.
 * These are the basic types that can be serialized to JSON.
 */
export type JsonPrimitive = string | number | boolean | null;

/**
 * Value types allowed in provider results.
 *
 * This type accepts:
 * - Primitive JSON values (string, number, boolean, null, undefined)
 * - Arrays of values
 * - Any object (Record<string, unknown>)
 *
 * The broad object type (Record<string, unknown>) ensures that domain types
 * like Memory[], Character, Content, etc. are accepted without requiring
 * unsafe 'as unknown as' casts, while still maintaining JSON-serializable
 * semantics at runtime.
 */
export type ProviderValue =
	| JsonPrimitive
	| JsonValue
	| Uint8Array
	| bigint
	| object
	| ProviderValue[]
	| { [key: string]: ProviderValue | undefined }
	| undefined;

/**
 * Data record type that accepts any JSON-serializable values.
 * This is broader than ProviderValue to accommodate domain types
 * like Memory[], Character, Content without requiring casts.
 * The index signature allows dynamic property access.
 */
export type ProviderDataRecord = {
	[key: string]: ProviderValue;
};

/**
 * Result returned by a provider
 */
export interface ProviderResult {
	/** Human-readable text for LLM prompt inclusion */
	text?: string;

	/** Key-value pairs for template variable substitution */
	values?: Record<string, ProviderValue>;

	/**
	 * Structured data for programmatic access by other components.
	 * Accepts JSON-serializable values and domain objects.
	 */
	data?: ProviderDataRecord;
}

/**
 * Provider for external data/services
 */
export interface Provider {
	/** Provider name */
	name: string;

	/** Description of the provider */
	description?: string;

	/** Compressed description for prompt-optimized rendering */
	descriptionCompressed?: string;

	/** Whether the provider is dynamic */
	dynamic?: boolean;

	/** Position of the provider in the provider list, positive or negative */
	position?: number;

	/**
	 * Whether the provider is private
	 *
	 * Private providers are not displayed in the regular provider list, they have to be called explicitly
	 */
	private?: boolean;

	/** Keywords used to determine relevance for action filtering */
	relevanceKeywords?: string[];

	/**
	 * Domain contexts this provider belongs to.
	 * The context-routing classifier uses these to decide which providers to
	 * include in the planner's state composition for a given turn.
	 */
	contexts?: AgentContext[];

	/**
	 * Additional providers that should run alongside this provider when it is
	 * selected by the planner. Use this for provider composition, not semantic
	 * routing.
	 */
	companionProviders?: string[];

	/** Data retrieval function */
	get: (
		runtime: IAgentRuntime,
		message: Memory,
		state: State,
	) => Promise<ProviderResult>;
}

/**
 * Result returned by an action after execution
 * Used for action chaining and state management
 */
export interface ActionResult {
	/** Whether the action succeeded */
	success: boolean;

	/** Optional text description of the result */
	text?: string;

	/** Values to merge into the state */
	values?: Record<string, ProviderValue>;

	/**
	 * Data payload containing action-specific results.
	 * Accepts any JSON-serializable object values including domain types.
	 */
	data?: ProviderDataRecord;

	/** Error information if the action failed */
	error?: string | Error;

	/** Whether to continue the action chain (for chained actions) */
	continueChain?: boolean;

	/** Optional cleanup function to execute after action completion */
	cleanup?: () => void | Promise<void>;
}

/**
 * Context provided to actions during execution
 * Allows actions to access previous results and execution state
 */
export interface ActionContext {
	/** Results from previously executed actions in this run */
	previousResults: ActionResult[];

	/** Get a specific previous result by action name */
	getPreviousResult?: (actionName: string) => ActionResult | undefined;
}

/**
 * Canonical callback type for streaming response chunks.
 *
 * WHY one type: Before this consolidation the same `(chunk, messageId?) => …`
 * signature was inlined in 8+ locations across runtime, model, message-service,
 * and streaming-context types — with inconsistent return types (`Promise<void>`
 * vs `void | Promise<void>`). Adding data (e.g. `accumulated`) required editing
 * every copy. A single alias eliminates drift and makes future extensions
 * (field name, token index, session handle) a one-line additive change.
 *
 * WHY `accumulated`: Two independent XML stream extractors in `useModel`
 * previously caused TTS garbling because consumers had to re-derive the full
 * text from deltas — and the two extractors produced deltas at different
 * timings. Providing the authoritative accumulated text from the extractor
 * makes that entire category of reassembly bugs impossible.
 *
 * WHY `void | Promise<void>`: The most permissive return — allows both sync
 * callbacks (simple loggers, test spies) and async ones (network, TTS).
 *
 * @param chunk - Delta text since the last emission for this field.
 * @param messageId - Streaming session / message identifier (UUID or opaque string).
 * @param accumulated - Full extracted text so far for the streaming field.
 *   Present when the emission originates from a ValidationStreamExtractor
 *   (structured XML output). Undefined for raw-token streams (useModel
 *   without an extractor) where no field-level accumulation exists.
 */
export type StreamChunkCallback = (
	chunk: string,
	messageId?: string,
	accumulated?: string,
) => void | Promise<void>;

/**
 * Options passed to action handlers during execution
 * Provides context about the current execution and multi-step plans
 */
export interface HandlerOptions {
	/** Context with previous action results and utilities */
	actionContext?: ActionContext;

	/** Multi-step action plan information */
	actionPlan?: ActionPlan;

	/** Optional stream chunk callback for streaming responses */
	onStreamChunk?: StreamChunkCallback;

	/**
	 * Validated input parameters extracted from the conversation.
	 * Only present when the action defines parameters and they were successfully extracted.
	 *
	 * Parameters are validated against the action's parameter schema before being passed here.
	 * Optional parameters may be undefined if not provided in the conversation.
	 *
	 * @example
	 * ```typescript
	 * handler: async (runtime, message, state, options) => {
	 *   const params = options?.parameters;
	 *   if (params) {
	 *     const targetUser = params.targetUser as string;
	 *     const platform = params.platform as string ?? "telegram"; // backfill default
	 *   }
	 * }
	 * ```
	 */
	parameters?: ActionParameters;

	/**
	 * Parameter validation errors, if the action defined parameters but extraction/validation was incomplete.
	 *
	 * Actions SHOULD handle these errors gracefully (e.g. ask the user for missing required values,
	 * or infer from context when safe).
	 */
	parameterErrors?: string[];

	/** Allow extensions from plugins */
	[key: string]: JsonValue | object | undefined;
}

File: packages/rust/src/types/components.rs
//! Component types (proto-backed) with runtime helpers.
//!
//! This module defines the core component types for actions, providers, and evaluators,
//! along with their handlers and runtime definitions.

use async_trait::async_trait;
use serde::{Deserialize, Serialize};
use serde_json::Value as JsonValue;
use std::collections::HashMap;
use std::future::Future;
use std::pin::Pin;

pub use super::generated::eliza::v1::{
    ActionExample, ActionParameter, ActionParameterSchema, EvaluationExample,
};
use super::memory::Memory;
use super::primitives::Content;
use super::state::State;

/// Type alias for action parameters stored as a JSON key-value map.
pub type ActionParameters = HashMap<String, JsonValue>;

/// Context passed to action handlers containing information about prior action executions.
#[derive(Clone, Debug, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ActionContext {
    /// Results from previously executed actions in the current chain.
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub previous_results: Vec<ActionResult>,
}

/// Options passed to action, provider, and evaluator handlers.
#[derive(Clone, Debug, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct HandlerOptions {
    /// Context from prior action executions.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub action_context: Option<ActionContext>,
    /// Serialized action plan JSON for multi-step executions.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub action_plan_json: Option<String>,
    /// Parameters passed to the handler.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub parameters: Option<ActionParameters>,
}

/// Result returned by action handlers after execution.
#[derive(Clone, Debug, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ActionResult {
    /// Whether the action completed successfully.
    pub success: bool,
    /// Optional text response from the action.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub text: Option<String>,
    /// Optional key-value pairs to merge into state values.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub values: Option<ActionParameters>,
    /// Optional structured data returned by the action.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub data: Option<ActionParameters>,
    /// Error message if the action failed.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

/// Result returned by provider handlers after execution.
#[derive(Clone, Debug, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ProviderResult {
    /// Optional text content provided to the context.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub text: Option<String>,
    /// Optional key-value pairs to merge into state values.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub values: Option<ActionParameters>,
    /// Optional structured data from the provider.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub data: Option<ActionParameters>,
}

/// Helper to insert a value into an optional HashMap, creating it if needed.
fn insert_map_value(target: &mut Option<ActionParameters>, key: String, value: JsonValue) {
    let map = target.get_or_insert_with(HashMap::new);
    map.insert(key, value);
}

impl ActionResult {
    /// Creates a successful action result with the given message.
    pub fn success(message: impl Into<String>) -> Self {
        ActionResult {
            success: true,
            text: Some(message.into()),
            values: None,
            data: None,
            error: None,
        }
    }

    /// Creates a successful action result with the given text message.
    pub fn success_with_text(message: &str) -> Self {
        ActionResult {
            success: true,
            text: Some(message.to_string()),
            values: None,
            data: None,
            error: None,
        }
    }

    /// Creates a failed action result with the given error message.
    pub fn failure(message: &str) -> Self {
        ActionResult {
            success: false,
            text: None,
            values: None,
            data: None,
            error: Some(message.to_string()),
        }
    }

    /// Adds a key-value pair to the result's values map.
    pub fn with_value(mut self, key: impl Into<String>, value: impl Into<JsonValue>) -> Self {
        insert_map_value(&mut self.values, key.into(), value.into());
        self
    }

    /// Adds a key-value pair to the result's data map.
    pub fn with_data(mut self, key: impl Into<String>, value: impl Into<JsonValue>) -> Self {
        insert_map_value(&mut self.data, key.into(), value.into());
        self
    }
}

impl ProviderResult {
    /// Creates a new provider result with the given text.
    pub fn new(text: impl Into<String>) -> Self {
        ProviderResult {
            text: Some(text.into()),
            values: None,
            data: None,
        }
    }

    /// Creates a provider result with the given text content.
    pub fn with_text(text: impl Into<String>) -> Self {
        ProviderResult {
            text: Some(text.into()),
            values: None,
            data: None,
        }
    }

    /// Creates an empty provider result with no content.
    pub fn empty() -> Self {
        ProviderResult::new("")
    }

    /// Adds a key-value pair to the result's values map.
    pub fn with_value(mut self, key: impl Into<String>, value: impl Into<JsonValue>) -> Self {
        insert_map_value(&mut self.values, key.into(), value.into());
        self
    }

    /// Adds a key-value pair to the result's data map.
    pub fn with_data(mut self, key: impl Into<String>, value: impl Into<JsonValue>) -> Self {
        insert_map_value(&mut self.data, key.into(), value.into());
        self
    }
}

// Runtime definitions (not in proto)

/// Definition metadata for an action.
#[derive(Clone, Debug, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", bound(serialize = "", deserialize = ""))]
pub struct ActionDefinition {
    /// Unique name identifier for the action.
    pub name: String,
    /// Human-readable description of what the action does.
    pub description: String,
    /// Alternative names that can trigger this action.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub similes: Option<Vec<String>>,
    /// Examples are skipped during serialization as they contain proto types
    #[serde(skip, default)]
    pub examples: Option<Vec<Vec<ActionExample>>>,
    /// Execution priority (higher runs first).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub priority: Option<i32>,
    /// Tags for categorization and filtering.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tags: Option<Vec<String>>,
    /// Parameters are skipped during serialization as they contain proto types
    #[serde(skip, default)]
    pub parameters: Option<Vec<ActionParameter>>,
}

/// Definition metadata for a provider.
#[derive(Clone, Debug, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ProviderDefinition {
    /// Unique name identifier for the provider.
    pub name: String,
    /// Human-readable description of what the provider provides.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    /// Whether the provider generates dynamic content.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub dynamic: Option<bool>,
    /// Position in provider ordering (lower runs first).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub position: Option<i32>,
    /// Whether the provider output is private (not included in prompts).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub private: Option<bool>,
}

/// Definition metadata for an evaluator.
#[derive(Clone, Debug, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", bound(serialize = "", deserialize = ""))]
pub struct EvaluatorDefinition {
    /// Unique name identifier for the evaluator.
    pub name: String,
    /// Human-readable description of what the evaluator does.
    pub description: String,
    /// Whether to run this evaluator on every message.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub always_run: Option<bool>,
    /// Alternative names that can trigger this evaluator.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub similes: Option<Vec<String>>,
    /// Examples are skipped during serialization as they contain proto types
    #[serde(skip, default)]
    pub examples: Vec<EvaluationExample>,
}

/// Callback type for handling content and producing memories.
pub type HandlerCallback = Box<
    dyn Fn(Content) -> Pin<Box<dyn Future<Output = Vec<Memory>> + Send + 'static>> + Send + Sync,
>;

/// Callback type for handling streaming text chunks.
pub type StreamChunkCallback =
    Box<dyn Fn(&str, Option<&str>) -> Pin<Box<dyn Future<Output = ()> + Send>> + Send + Sync>;

/// Trait for implementing action handlers.
///
/// Actions are discrete operations that the agent can perform in response to messages,
/// such as sending replies, executing commands, or interacting with external services.
#[async_trait]
pub trait ActionHandler: Send + Sync {
    /// Returns the definition metadata for this action.
    fn definition(&self) -> ActionDefinition;
    /// Validates whether this action should run for the given message and state.
    async fn validate(&self, message: &Memory, state: Option<&State>) -> bool;
    /// Executes the action and returns the result.
    async fn handle(
        &self,
        message: &Memory,
        state: Option<&State>,
        options: Option<&HandlerOptions>,
    ) -> Result<Option<ActionResult>, anyhow::Error>;
}

/// Trait for implementing provider handlers.
///
/// Providers supply contextual information to the agent's decision-making process,
/// such as time, character details, recent messages, or external data.
#[async_trait]
pub trait ProviderHandler: Send + Sync {
    /// Returns the definition metadata for this provider.
    fn definition(&self) -> ProviderDefinition;
    /// Retrieves the provider's content for the given message and state.
    async fn get(&self, message: &Memory, state: &State) -> Result<ProviderResult, anyhow::Error>;
}

/// Trait for implementing evaluator handlers.
///
/// Evaluators analyze messages and state to determine conditions or extract information,
/// such as detecting sentiment, identifying intents, or triggering follow-up actions.
#[async_trait]
pub trait EvaluatorHandler: Send + Sync {
    /// Returns the definition metadata for this evaluator.
    fn definition(&self) -> EvaluatorDefinition;
    /// Validates whether this evaluator should run for the given message and state.
    async fn validate(&self, message: &Memory, state: Option<&State>) -> bool;
    /// Executes the evaluation and returns the result.
    async fn handle(
        &self,
        message: &Memory,
        state: Option<&State>,
        options: Option<&HandlerOptions>,
    ) -> Result<Option<ActionResult>, anyhow::Error>;
}
